### ISSUE TITLE
Develop CosmosDbMigration package for console applications which applied to NoSQL

### ISSUE THREAD

**Issue by kanilsz** [Marusyk/MSA.BuildingBlocks#10]
It would be great to execute Cosmos db migrations with console application.
Necessary to provide basic operations:

1. On database 

- [ ]  Delete container
- [ ]  Recreate container  with new partition key
- [ ]  Clone container
- [ ]  Add indexing policy
- [ ]  Create container
- [ ]  Replace indexing policy

2. On container

- [ ]  Add property to item\items on any nesting level
- [ ]  Remove property from item\items on any nesting level
- [ ]  Upsert item\items
- [ ]  Remove item\items by query

Additionally could be added some console logs of made operation and it's RUs cost.

**Comment by Marusyk** [comment]
great idea. let's start doing it one by one.

**Comment by Marusyk** [comment]
One more requirement. Currently, the migration process starts on the startup of the application. We could have another option - run it as a sidecar. I'll try to explain what I mean.

There is sidecar pattern: https://www.baeldung.com/ops/kubernetes-pods-sidecar-containers
where you can run 1+ containers for k8s deployment. It could fix the issue with Rece Conditions (which we fixed with @kanilsz  a long time ago).
Sidecar init container runs before starting our application instances.
![image](https://github.com/Marusyk/MSA.BuildingBlocks/assets/9221694/fc00f83a-ecb8-4516-9de0-6ff2a53cb3c0)

and there are 2 strategies for implementation:
1. **The separated docker image for containers** - very interesting approach but has extra complexity and CI pipeline is required for such migration-image
2. **Run the same application with arguments**

**Comment by kanilsz** [comment]
I'm ready to try and implement it. Could we make it in the scope of another feature/issue? For now, I want to end with basic functionality. Moreover, I have some thoughts on how it could be improved. What do you think? As for me, it's better to do it in 
the different scopes.

**Comment by Marusyk** [comment]
glad to hear that 😉
I'd propose to have 1 generic issue like this with the name "[package name] package" and use it as an epic to define the requirements. and then the more specific issues can be created for each requirement or unit of work (1PR = 1 issue)

**Comment by Marusyk** [comment]
There is a Migration tool for MongoDb and I see it is very similar to what we have. I like our approach even more
https://bitbucket.org/i_am_a_kernel/mongodbmigrations/src/master/

**Comment by kanilsz** [comment]
> There is a Migration tool for MongoDb and I see it is very similar to what we have. I like our approach even more https://bitbucket.org/i_am_a_kernel/mongodbmigrations/src/master/

Our you mean here?) Or in previous common packages) Also I unlinked the pr for this issue because seem it is going to close after merge.

**Comment by kanilsz** [comment]
> one more: instead of `ARepository.Init(database); BRepository.Init(database); CRepository.Init(database);` we could have a factory patter implemented.

Yes, it could be. I have more thoughts about something like building pattern for migrations. I see it as:
(For all items) MakeMigrations(items)

**Comment by kanilsz** [comment]
Cosmos DB possibly could have more than a million records. So for an executor it will be a loooong time to execute the migrations. Moreover? it could be crashed. 
What I propose:

1. Provide migration by some chunk (1000 items)
2. If it falls we tell to execute than the items 600 is failed.
3. Let him fix it by himself.
4. Start the execution from 600.

For me it looks very hard to implement by I think use case is real.

**Comment by kanilsz** [comment]
It would be great to have integration tests that will validate and ensure that the migration was made successfully. 

There are two possible options that I know.
1. TestContainers https://testcontainers.com/modules/cosmodb/
2. Up the cosmos db container by ourselves (not sure that the issue with certificates that we have is resolved)

Probably some unit tests for a future builder will ensure that operations were made successfully.
